README.md
# [기술 블로그](https://jeongjoohong.github.io/project/)

[![License](https://img.shields.io/badge/license-MIT-lightgrey.svg)](https://raw.githubusercontent.com/jeongjoohong/project/master/LICENSE)

Jekyll을 이용하여 생성된 블로그의 저장소 입니다. 프로젝트의 블로그를 작성하는데 사용하고 있습니다.
변경기록은 [CHANGELOG.md](https://github.com/jeongjoohong/project/blob/gh-pages/CHANGELOG.md)를 통해 확인할 수 있습니다.

---

## Features

- Compatible with Jekyll 3.x and GitHub Pages
- Several responsive layout options (single, archive index, splash, and paginated home page)
- UI localized text in Korean

## Pages

| Name    | Description |
| ------- | ----------- |
| [home](https://jeongjoohong.github.io/project/) | Paginated home page. |
| [category](https://jeongjoohong.github.io/project/categories/) | Posts grouped by category. |
| [tag](https://jeongjoohong.github.io/project/tags/) | Posts grouped by tags. |


## Structure

```
project
├── _data                      # data files for customizing the theme
|  ├── navigations.yml         # main navigation links
|  └── ui-text.yml             # text used throughout the theme's UI
├── _includes
|  ├── analytics-providers     # snippets for analytics (Google and custom)
|  ├── comments-providers      # snippets for comments (Disqus, Facebook, Google+, and custom)
|  ├── footer                  # custom snippets to add to site footer
|  ├── head                    # custom snippets to add to site head
|  ├── base_path               # site.url + site.baseurl shortcut
|  ├── feature_row             # feature row helper
|  ├── gallery                 # image gallery helper
|  ├── group-by-array          # group by array helper for archives
|  ├── nav_list                # navigation list helper
|  ├── toc                     # table of contents helper
|  └── ...
├── _layouts
|  ├── archive-taxonomy.html   # tag/category archive for Jekyll Archives plugin
|  ├── archive.html            # archive listing documents in an array
|  ├── compress.html           # compresses HTML in pure Liquid
|  ├── default.html            # base for all other layouts
|  ├── home.html               # home page
|  ├── single.html             # single document (post/page/etc)
|  └── splash.html             # splash page
├── _sass                      # SCSS partials
├── assets
|  ├── css
|  |  └── main.scss            # main stylesheet, loads SCSS partials from _sass
|  ├── fonts
|  |  └── fontawesome-webfont  # Font Awesome webfonts
|  ├── images                  # image assets for posts/pages/collections/etc.
|  ├── js
|  |  ├── plugins              # jQuery plugins
|  |  ├── vendor               # vendor scripts
|  |  ├── _main.js             # plugin settings and other scripts to load after jQuery
|  |  └── main.min.js          # optimized and concatenated script file loaded before </body>
├── _config.yml                # site configuration
├── Gemfile                    # gem file dependencies
├── index.html                 # paginated home page showing recent posts
└── package.json               # NPM build scripts
```

## Stylesheets

```
project
├── _sass
|  ├── vendor               # vendor SCSS partials
|  |   ├── breakpoint       # media query mixins
|  |   ├── font-awesome     # Font Awesome icons
|  |   ├── magnific-popup   # Magnific Popup lightbox
|  |   └── susy             # Susy grid system
|  ├── _animations.scss     # animations
|  ├── _archive.scss        # archives (list, grid, feature views)
|  ├── _base.scss           # base HTML elements
|  ├── _buttons.scss        # buttons
|  ├── _footer.scss         # footer
|  ├── _masthead.scss       # masthead
|  ├── _mixins.scss         # mixins (em function, clearfix)
|  ├── _navigation.scss     # nav links (breadcrumb, priority+, toc, pagination, etc.)
|  ├── _notices.scss        # notices
|  ├── _page.scss           # pages
|  ├── _print.scss          # print styles
|  ├── _reset.scss          # reset
|  ├── _sidebar.scss        # sidebar
|  ├── _syntax.scss         # syntax highlighting
|  ├── _tables.scss         # tables
|  ├── _utilities.scss      # utility classes (text/image alignment)
|  └── _variables.scss      # theme defaults (fonts, colors, etc.)
├── assets
|  ├── css
|  |  └── main.scss         # main stylesheet, loads SCSS partials in _sass
```

## JavaScript

```
project
├── assets
|  ├── js
|  |  ├── plugins
|  |  |   ├── jquery.fitvids.js            # fluid width video embeds
|  |  |   ├── jquery.greedy-navigation.js  # priority plus navigation
|  |  |   ├── jquery.magnific-popup.js     # responsive lightbox
|  |  |   └── jquery.smooth-scroll.min.js  # make same-page links scroll smoothly
|  |  ├── vendor
|  |  |   └── jquery
|  |  |       └── jquery-1.12.1.min.js
|  |  ├── _main.js                         # jQuery plugin settings and other scripts
|  |  └── main.min.js                      # concatenated and minified scripts
```

---

## Credits

- [Jekyll](http://jekyllrb.com/)
- [Minimal Mistakes Jekyll Theme](https://mmistakes.github.io/minimal-mistakes/)

## License

See the [LICENSE](https://github.com/jeongjoohong/project/blob/gh-pages/LICENSE) file.


CHANGELOG.md
## v1.5

### 관리

* 테마 업데이트

  * Minimal Mistakes 테마(v4.5.0)


## v1.4

### 관리

* 테마 업데이트

  * Minimal Mistakes 테마(v4.4.1)


## v1.3

### 관리

* 테마 업데이트

  * Minimal Mistakes 테마(v4.4.0)


## v1.2

### 관리

* 테마 업데이트

  * 프린트 스타일 변경

  * 라틴어 추가

  * 비디오 전체화면 오류 수정


## v1.1

### 관리

* 테마 업데이트

  * Minimal Mistakes 테마(v4.3.1)


## v1.0

### 관리

* 사이트 생성


Test/test_convolution.m
function w = test_convolution(u, v, option)
    w = zeros(1, length(u) + length(v) - 1);
    for i = 1:length(u)+length(v)-1
        index = i;
        tmp = 0;
        for j = 1:length(v)
            if (index >= 1) && ~(index > length(u))
                tmp = tmp + u(index) * v(j);
            end
            index = index - 1;
            w(i) = tmp;
        end
    end
    if strcmp(option, 'same')
        start = ceil((length(w) - length(u)) / 2);
        w = w(start+1:start+length(u));
    end
end

SignalProcessing/decision_test.m
close all;
clear;
clc;

bandw = 0.1;
freq1 = 50;
freq2 = 120;
amp1 = 0.6;
amp2 = 0.8;
plotCol = 7;
plotInd = 0;

Fs = 400;       % Sampling frequency
T = 1/Fs;       % Sampling period
L = 400;        % Length of signal
t = (0:L-1)*T;  % Time vector

S = amp1*sin(2*pi*50*t)+amp2*sin(2*pi*120*t);
X = S+randn(size(t));

figure('units','normalized','outerposition',[0 0 1 1]);
plotInd = plotInd+1;
subplot(plotCol,2,plotInd);
plot(1000*t,X);
title('Signal with Noise');
xlabel('t (milliseconds)');
ylabel('X(t)');

Y = fftshift(fft(X,L));
fY = (-numel(Y)/2:numel(Y)/2-1);
plotInd = plotInd+1;
subplot(plotCol,2,plotInd);
plot(fY, abs(Y));
title('FFT of X(t)');
xlabel('f (Hz)');
ylabel('|X(f)|');

A1 = sinc(bandw*(t-0.5)) .* cos(2*pi*freq1*t);
plotInd = plotInd+1;
subplot(plotCol,2,plotInd);
plot(1000*t,A1);
title(sprintf('Sinc (%.1fHz, centered, shifted %dHz)',bandw,freq1));
xlabel('t (milliseconds)');
ylabel(sprintf('sinc(%.1f*(t-0.5)) * cos(2*pi*%d*t)',bandw,freq1));

A2 = sinc(bandw*(t-0.5)) .* cos(2*pi*freq2*t);
plotInd = plotInd+1;
subplot(plotCol,2,plotInd);
plot(1000*t,A2);
title(sprintf('Sinc (%.1fHz, centered, shifted %dHz)',bandw,freq2));
xlabel('t (milliseconds)');
ylabel(sprintf('sinc(%.1f*(t-0.5)) * cos(2*pi*%d*t)',bandw,freq2));

B1 = fftshift(fft(A1,L));
fB1 = (-numel(B1)/2:numel(B1)/2-1);
plotInd = plotInd+1;
subplot(plotCol,2,plotInd);
plot(fB1, abs(B1));
title('FFT of sinc(t)');
xlabel('f (Hz)');

B2 = fftshift(fft(A2,L));
fB2 = (-numel(B2)/2:numel(B2)/2-1);
plotInd = plotInd+1;
subplot(plotCol,2,plotInd);
plot(fB2, abs(B2));
title('FFT of sinc(t)');
xlabel('f (Hz)');

C1 = ifft(ifftshift(B1),L);
plotInd = plotInd+1;
subplot(plotCol,2,plotInd);
plot(1000*t,C1);
title('IFFT of sinc(f)');
xlabel('t (milliseconds)');

C2 = ifft(ifftshift(B2),L);
plotInd = plotInd+1;
subplot(plotCol,2,plotInd);
plot(1000*t,C2);
title('IFFT of sinc(f)');
xlabel('t (milliseconds)');

D1 = conv(X,C1);
plotInd = plotInd+1;
subplot(plotCol,2,plotInd);
plot(D1);
title('Convolution');

D2 = conv(X,C2);
plotInd = plotInd+1;
subplot(plotCol,2,plotInd);
plot(D2);
title('Convolution');

D1 = conv(X,C1);
plotInd = plotInd+1;
subplot(plotCol,2,plotInd);
plot(abs(D1));
title('Absolute of Convolution');

D2 = conv(X,C2);
plotInd = plotInd+1;
subplot(plotCol,2,plotInd);
plot(abs(D2));
title('Absolute of Convolution');

E = [max(abs(D1)) max(abs(D2))];
plotInd = plotInd+1;
subplot(plotCol,1,ceil(plotInd/2));
hold on
for i = 1 : length(E)
    b(i) = bar(i, E(i));
end
hold off
title(sprintf('Maximum Absolute Value of Convolution %dHz, and %dHz',freq1,freq2));
ylabel('max(|conv(X(t), sinc(t))|)');
set(gca, 'XTickLabel',E,'XTick',1:length(E))
if max(abs(D1)) > max(abs(D2))
    b(1).LineWidth = 2;
    b(1).EdgeColor = 'red';
else
    b(2).LineWidth = 2;
    b(2).EdgeColor = 'red';
end


SignalProcessing/test_decision.m
close all;
clear;
clc;

%% Variables
Ls = 400;           % Length of signal
Lf = Ls;            % Length of filter
Fs = 100;           % Sampling frequency
T = 1 / Fs;         % Sampling period
ts = (0:Ls-1) * T;  % Time vector of signal
tf = (0:Lf-1) * T;  % Time vector of filter

Freq1 = 7;
Freq2 = 8;
Amp1 = 0.7;
Amp2 = 0.6;
% AmpN = 0.01;
bandw = 1;
freq1 = Freq1;
freq2 = Freq2;

plotCol = 5;
plotInd = 0;

%% Signal generation
S1 = Amp1 * sin(2 * pi * Freq1 * ts);
S2 = Amp2 * sin(2 * pi * Freq2 * ts);
S = S1 + S2;
% X = S + AmpN * randn(size(ts));
X = S;

figure('units', 'normalized', 'outerposition', [0 0 1 1]);
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
plot(ts * 1e3, X);
title(sprintf('Signal : %.1f*sin(2*pi*%d*t) + %.1f*sin(2*pi*%d*ts)', Amp1, Freq1, Amp2, Freq2));
xlabel('t (milliseconds)');
ylabel('X(t)');

%% FFT of received signal
Y = fftshift(fft(X(1:Lf)));
abscissa = linspace(-Fs / 2, Fs / 2 - Fs / Lf, length(Y));
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
plot(abscissa, abs(Y));
title('FFT of signal');
xlabel('f (Hz)');
ylabel('|X(f)|');

%% Sinc filters
A1 = sinc(bandw * (tf - (Lf/Fs/2))) .* cos(2 * pi * freq1 * tf);
A1 = A1 ./ sum(abs(A1) .^ 2);   % Normalization
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
plot(tf * 1e3, A1);
title(sprintf('Sinc1 : %.1fHz bandwidth, centered, shifted %dHz', bandw, freq1));
xlabel('t (milliseconds)');

A2 = sinc(bandw * (tf - (Lf/Fs/2))) .* cos(2 * pi * freq2 * tf);
A2 = A2 ./ sum(abs(A2) .^ 2);   % Normalization
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
plot(tf * 1e3, A2);
title(sprintf('Sinc2 : %.1fHz bandwidth, centered, shifted %dHz', bandw, freq2));
xlabel('t (milliseconds)');

%% Examination of sinc filters using FFT and IFFT
test1FFT = fftshift(fft(A1));
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
plot(abscissa, abs(test1FFT));
line([Freq1 Freq1 ; -Freq1 -Freq1]', [get(gca, 'YLim') ; get(gca, 'YLim')]', 'Color', 'red', 'LineStyle', '--');
title('FFT of sinc1');
xlabel('f (Hz)');

test2FFT = fftshift(fft(A2));
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
plot(abscissa, abs(test2FFT));
line([Freq2 Freq2 ; -Freq2 -Freq2]', [get(gca, 'YLim') ; get(gca, 'YLim')]', 'Color', 'red', 'LineStyle', '--');
title('FFT of sinc2');
xlabel('f (Hz)');

%% Convolution
B1 = conv(X(1:Lf), A1, 'same');
C1 = test_convolution(X(1:Lf), A1, 'same');
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
plot(tf * 1e3, B1, tf * 1e3, C1, '-.');
legend('matlab function', 'test function', 'Location', 'southeast');
title('Convolution of signal and sinc1');
xlabel('t (milliseconds)');

B2 = conv(X(1:Lf), A2, 'same');
C2 = test_convolution(X(1:Lf), A2, 'same');
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
plot(tf * 1e3, B2, tf * 1e3, C2, '-.');
legend('matlab function', 'test function', 'Location', 'southeast');
title('Convolution of signal and sinc2');
xlabel('t (milliseconds)');

%% Comparison of powers
Pwr_S1 = sum(S1(1:Lf) .^ 2);
Pwr_S2 = sum(S2(1:Lf) .^ 2);
Pwr_B1 = sum(B1 .^ 2);
Pwr_B2 = sum(B2 .^ 2);
Pwr_C1 = sum(C1 .^ 2);
Pwr_C2 = sum(C2 .^ 2);
Pwr = [Pwr_S1 Pwr_B1 Pwr_C1 ; Pwr_S2 Pwr_B2 Pwr_C2];
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
bar(Pwr);
set(gca, 'XTickLabel', {sprintf('%dHz', freq1), sprintf('%dHz', freq2)});
title('Comparison of signal powers');
legend('source', 'matlab function', 'test function', 'Location', 'southeast', 'Orientation', 'horizontal');

%% Loss powers of signal
LossPwr_B1 = (1 - Pwr_B1 / Pwr_S1) * 100;   % Loss of signal (%)
LossPwr_B2 = (1 - Pwr_B2 / Pwr_S2) * 100;   % Loss of signal (%)
LossPwr_C1 = (1 - Pwr_C1 / Pwr_S1) * 100;   % Loss of signal (%)
LossPwr_C2 = (1 - Pwr_C2 / Pwr_S2) * 100;   % Loss of signal (%)
Loss = [LossPwr_B1 LossPwr_C1 ; LossPwr_B2 LossPwr_C2];
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
bar(Loss);
set(gca, 'XTickLabel', {sprintf('%dHz', freq1), sprintf('%dHz', freq2)});
title('Loss powers of signal (%)');
legend('matlab function', 'test function');
ylabel('percentage');
ylim([0 100]);


BlinkerModule/OneEuroFilter.h
#ifndef OneEuroFilter_h
#define OneEuroFilter_h

#include "Arduino.h"
#include "LowPassFilter.h"

class OneEuroFilter {
  public:
    OneEuroFilter(double freq);
    OneEuroFilter(double freq, double mincutoff);
    OneEuroFilter(double freq, double mincutoff, double beta_);
    OneEuroFilter(double freq, double mincutoff, double beta_, double dcutoff);
    double filter(double value);
    double filter(double value, double timestamp);
  protected:
    double freq ;
    double mincutoff ;
    double beta_ ;
    double dcutoff ;
    LowPassFilter *x ;
    LowPassFilter *dx ;
    double lasttime ;
    double UndefinedTime = -1;
    void init(double freq, double mincutoff, double beta_, double dcutoff);
    double alpha(double cutoff);
    void setFrequency(double f);
    void setMinCutoff(double mc);
    void setBeta(double b);
    void setDerivateCutoff(double dc);
};

#endif



BlinkerModule/OneEuroFilter.cpp
#include "OneEuroFilter.h"

OneEuroFilter::OneEuroFilter(double freq) {
  init(freq, 1.0, 0.0, 1.0);
}

OneEuroFilter::OneEuroFilter(double freq, double mincutoff) {
  init(freq, mincutoff, 0.0, 1.0);
}

OneEuroFilter::OneEuroFilter(double freq, double mincutoff, double beta_) {
  init(freq, mincutoff, beta_, 1.0);
}

OneEuroFilter::OneEuroFilter(double freq, double mincutoff, double beta_, double dcutoff) {
  init(freq, mincutoff, beta_, dcutoff);
}

void OneEuroFilter::init(double freq, double mincutoff, double beta_, double dcutoff) {
  setFrequency(freq) ;
  setMinCutoff(mincutoff) ;
  setBeta(beta_) ;
  setDerivateCutoff(dcutoff) ;
  x = new LowPassFilter(alpha(mincutoff)) ;
  dx = new LowPassFilter(alpha(dcutoff)) ;
  lasttime = UndefinedTime ;
}

double OneEuroFilter::alpha(double cutoff) {
  double te = 1.0 / freq ;
  double tau = 1.0 / (2 * M_PI * cutoff) ;
  return 1.0 / (1.0 + tau / te) ;
}

void OneEuroFilter::setFrequency(double f) {
  if (f <= 0) {
    Serial.println("freq should be >0");
  }
  freq = f;
}

void OneEuroFilter::setMinCutoff(double mc) {
  if (mc <= 0) {
    Serial.println("mincutoff should be >0");
  }
  mincutoff = mc ;
}

void OneEuroFilter::setBeta(double b) {
  beta_ = b ;
}

void OneEuroFilter::setDerivateCutoff(double dc) {
  if (dc <= 0) {
    Serial.println("dcutoff should be >0");
  }
  dcutoff = dc;
}

double OneEuroFilter::filter(double value) {
  return filter(value, UndefinedTime);
}

double OneEuroFilter::filter(double value, double timestamp) {
  if (lasttime != UndefinedTime && timestamp != UndefinedTime) {
    freq = 1.0 / (timestamp - lasttime) ;
  }
  lasttime = timestamp ;
  double dvalue = x->hasLastRawValue() ? (value - x->lastRawValue()) * freq : 0.0; // FIXME: 0.0 or value?
  double edvalue = dx->filterWithAlpha(dvalue, alpha(dcutoff));
  double cutoff = mincutoff + beta_ * fabs(edvalue);
  return x->filterWithAlpha(value, alpha(cutoff));
}



SignalProcessing/LowPassFilter.cpp
#include "LowPassFilter.h"

LowPassFilter::LowPassFilter(double alpha) {
  init(alpha, 0);
}

LowPassFilter::LowPassFilter(double alpha, double initval) {
  init(alpha, initval);
}

void LowPassFilter::init(double alpha, double initval) {
  y = s = initval;
  setAlpha(alpha);
  initialized = false;
}

void LowPassFilter::setAlpha(double alpha) {
  if (alpha <= 0.0 || alpha > 1.0) {
      Serial.println("alpha should be in (0.0, 1.0)");
  }
  a = alpha;
}

double LowPassFilter::filter(double value) {
  double result ;
  if (initialized) {
    result = a * value + (1.0 - a) * s;
  } else {
    result = value;
    initialized = true;
  }
  y = value;
  s = result;
  return result;
}

double LowPassFilter::filterWithAlpha(double value, double alpha) {
  setAlpha(alpha);
  return filter(value);
}

bool LowPassFilter::hasLastRawValue(void) {
  return initialized;
}

double LowPassFilter::lastRawValue(void) {
  return y;
}



SignalProcessing/decision_test_byKang.m
close all;
clear;
clc;

%% Variables
Ls = 800;           % Length of signal
Lf = 300;           % Length of filter
Fs = 100;           % Sampling frequency
T = 1 / Fs;         % Sampling period
ts = (0:Ls-1) * T;  % Time vector of signal
tf = (0:Lf-1) * T;  % Time vector of filter

Freq1 = 7;
Freq2 = 8;
Amp1 = 0.7;
Amp2 = 0.6;
AmpN = 0.01;
bandw = 1;
freq1 = Freq1;
freq2 = Freq2;

plotCol = 5;
plotInd = 0;

%% Signal generation
S1 = Amp1 * sin(2 * pi * Freq1 * ts);
S2 = Amp2 * sin(2 * pi * Freq2 * ts);
S = S1 + S2;
% X = S + AmpN * randn(size(ts));
X = S;

figure('units', 'normalized', 'outerposition', [0 0 1 1]);
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
plot(ts * 1e3, X);
title(sprintf('Signal : %.1f*sin(2*pi*%d*t) + %.1f*sin(2*pi*%d*ts)', Amp1, Freq1, Amp2, Freq2));
xlabel('t (milliseconds)');
ylabel('X(t)');

%% FFT of received signal
Y = fftshift(fft(X(1:Lf), Lf));
f = linspace(-Fs / 2, Fs / 2 - Fs / Lf, length(Y));
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
plot(f, abs(Y));
title('FFT of signal');
xlabel('f (Hz)');
ylabel('|X(f)|');

%% Sinc filters
A1 = sinc(bandw * (tf - (Lf/Fs/2))) .* cos(2 * pi * freq1 * tf);
A1 = A1 ./ sum(abs(A1) .^ 2);   % Normalization
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
plot(tf * 1e3, A1);
title(sprintf('Sinc1 : %.1fHz bandwidth, centered, shifted %dHz', bandw, freq1));
xlabel('t (milliseconds)');

A2 = sinc(bandw * (tf - (Lf/Fs/2))) .* cos(2 * pi * freq2 * tf);
A2 = A2 ./ sum(abs(A2) .^ 2);   % Normalization
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
plot(tf * 1e3, A2);
title(sprintf('Sinc2 : %.1fHz bandwidth, centered, shifted %dHz', bandw, freq2));
xlabel('t (milliseconds)');

%% Examination of sinc filters using FFT and IFFT
test1FFT = fftshift(fft(A1, Lf));
% test1FVec = -numel(test1FFT)/2:numel(test1FFT)/2-1;
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
% plot(test1FVec, abs(test1FFT));
plot(f, abs(test1FFT));
line([Freq1 Freq1 ; -Freq1 -Freq1]', [get(gca, 'YLim') ; get(gca, 'YLim')]', 'Color', 'red', 'LineStyle', '--');
title('FFT of sinc1');
xlabel('f (Hz)');

test2FFT = fftshift(fft(A2, Lf));
% test2FVec = -numel(test2FFT)/2:numel(test2FFT)/2-1;
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
% plot(test2FVec, abs(test2FFT));
plot(f, abs(test2FFT));
line([Freq2 Freq2 ; -Freq2 -Freq2]', [get(gca, 'YLim') ; get(gca, 'YLim')]', 'Color', 'red', 'LineStyle', '--');
title('FFT of sinc2');
xlabel('f (Hz)');

%% Convolution
B1 = conv(X(1:Lf), A1, 'same');
C1 = convolution_test(X(1:Lf), A1, 'same');
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
plot(tf * 1e3, B1, tf * 1e3, C1, '-.');
legend('matlab function', 'test function', 'Location', 'southeast');
title('Convolution of signal and sinc1');
xlabel('t (milliseconds)');

B2 = conv(X(1:Lf), A2, 'same');
C2 = convolution_test(X(1:Lf), A2, 'same');
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
plot(tf * 1e3, B2, tf * 1e3, C2, '-.');
legend('matlab function', 'test function', 'Location', 'southeast');
title('Convolution of signal and sinc2');
xlabel('t (milliseconds)');

%% Comparison of powers
Pwr_S1 = sum(S1(1:Lf) .^ 2);
Pwr_S2 = sum(S2(1:Lf) .^ 2);
Pwr_B1 = sum(B1 .^ 2);
Pwr_B2 = sum(B2 .^ 2);
Pwr_C1 = sum(C1 .^ 2);
Pwr_C2 = sum(C2 .^ 2);
Pwr = [Pwr_S1 Pwr_B1 Pwr_C1 ; Pwr_S2 Pwr_B2 Pwr_C2];
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
bar(Pwr);
set(gca, 'XTickLabel', {sprintf('%dHz', freq1), sprintf('%dHz', freq2)});
title('Comparison of signal powers');
legend('source', 'matlab function', 'test function', 'Location', 'southeast');

%% Loss powers of signal
LossPwr_B1 = (1 - Pwr_B1 / Pwr_S1) * 100;   % Loss of signal (%)
LossPwr_B2 = (1 - Pwr_B2 / Pwr_S2) * 100;   % Loss of signal (%)
LossPwr_C1 = (1 - Pwr_C1 / Pwr_S1) * 100;   % Loss of signal (%)
LossPwr_C2 = (1 - Pwr_C2 / Pwr_S2) * 100;   % Loss of signal (%)
Loss = [LossPwr_B1 LossPwr_C1 ; LossPwr_B2 LossPwr_C2];
plotInd = plotInd + 1;
subplot(plotCol, 2, plotInd);
bar(Loss);
set(gca, 'XTickLabel', {sprintf('%dHz', freq1), sprintf('%dHz', freq2)});
title('Loss powers of signal (%)');
legend('matlab function', 'test function');
ylabel('percentage');
ylim([0 100]);
